forest/benchmarking/operator_estimation.py

```diff
@@ -1,557 +1,1210 @@
+
+    :param qc: a quantum computer object on which to run each program
+    :param programs: a list of programs to run
+    :param meas_qubits: groups of qubits to measure for each program
+    :param num_shots: the number of shots to run for each program
+    :return: a len(programs) long list of num_shots by num_meas_qubits bit arrays of results for
+        each program.
+    """
+    assert len(programs) == len(meas_qubits), 'The qubits to measure must be specified for each ' \
+                                              'program, one list of qubits per program.'
+
+    results = []
+    for program, qubits in zip(programs, meas_qubits):
+        if len(qubits) == 0:
+            # corresponds to measuring identity; no program needs to be run.
+            results.append(np.array([[]]))
+            continue
+        # copy the program so the original is not mutated
+        prog = program.copy()
+        ro = prog.declare('ro', 'BIT', len(qubits))
+        for idx, q in enumerate(qubits):
+            prog += MEASURE(q, ro[idx])
+
+        prog.wrap_in_numshots_loop(num_shots)
+        if use_compiler:
+            prog = qc.quil_to_native_quil(prog)
+        exe = qc.compiler.native_quil_to_executable(prog)
+        shots = qc.run(exe)
+        results.append(shots)
+    return results
+
+
+def shots_to_obs_moments(bitarray: np.ndarray, qubits: List[int], observable: PauliTerm,
+                         use_beta_dist_unbiased_prior: bool = False) -> Tuple[float, float]:
+    """
+    Calculate the mean and variance of the given observable based on the bitarray of results.
+
+    :param bitarray: results from running `qc.run`, a 2D num_shots by num_qubits array.
+    :param qubits: list of qubits in order corresponding to the bitarray results.
+    :param observable: the observable whose moments are calculated from the shot data
+    :param use_beta_dist_unbiased_prior: if true then the mean and variance are estimated from a
+        beta distribution that incorporates an unbiased Bayes prior. This precludes var = 0.
+    :return: tuple specifying (mean, variance)
+    """
+    coeff = complex(observable.coefficient)
+    if not np.isclose(coeff.imag, 0):
+        raise ValueError(f"The coefficient of an observable should not be complex.")
+    coeff = coeff.real
+
+    obs_qubits = [q for q, _ in observable]
+    # Identify classical register indices to select
+    idxs = [idx for idx, q in enumerate(qubits) if q in obs_qubits]
+
+    if len(idxs) == 0: # identity term
+        return coeff, 0
+
+    assert bitarray.shape[1] == len(qubits), 'qubits should label each column of the bitarray'
+
+    # Pick columns corresponding to qubits with a non-identity out_operation
+    obs_strings = bitarray[:, idxs]
+    # Transform bits to eigenvalues; ie (+1, -1)
+    my_obs_strings = 1 - 2 * obs_strings
+    # Multiply row-wise to get operator values.
+    obs_vals = np.prod(my_obs_strings, axis=1)
+
+    if use_beta_dist_unbiased_prior:
```

Comment:
So I thought about this for a bit (the reasoning also applies to the frequentist version below.

suppose the true value is  `p=1e-N` for some positive N. We will roughly need num_shots proportional to 1/p in order to get the correct accuracy.